lib/rules/no-restricted-syntax.js

```diff
@@ -18,18 +18,43 @@ module.exports = {
 
         schema: {
             type: "array",
-            items: [{ type: "string" }],
+            items: [{
+                oneOf: [
+                    {
+                        type: "string"
+                    },
+                    {
+                        type: "object",
+                        properties: {
+                            selector: { type: "string" },
+                            message: { type: "string" }
+                        },
+                        required: ["selector"],
+                        additionalProperties: false
+                    }
+                ]
+            }],
             uniqueItems: true,
             minItems: 0
         }
     },
 
     create(context) {
-        return context.options.reduce((result, selector) => Object.assign(result, {
-            [selector](node) {
-                context.report({ node, message: "Using '{{selector}}' is not allowed.", data: { selector } });
-            }
-        }), {});
+        return context.options.reduce((result, selectorOrObject) => {
+            const isStringFormat = (typeof selectorOrObject === "string");
+            const selector = isStringFormat ? selectorOrObject : selectorOrObject.selector;
+            const hasCustomMessage = !isStringFormat && Boolean(selectorOrObject.message);
+
+            return Object.assign(result, {
+                [selector](node) {
+                    context.report({
+                        node,
+                        message: hasCustomMessage ? selectorOrObject.message : "Using '{{selector}}' is not allowed.",
+                        data: { selector }
```

Comment:
If a custom message contains the substring `{{selector}}`, this will cause it to get replaced with the current selector. I doubt this will be a common problem in practice, but do you think it's worth documenting this or updating the logic to avoid it?